File: Operating_Systems/CPU-Scheduling/PriorityScheduling.js
// implement non-preemptive priority scheduling algorithm
// The non-preemptive priority scheduling algorithm is designed to select the process with the highest priority for execution. The process with the highest priority is selected from the ready queue and allocated the CPU. The priority of a process is generally an integer value. The process with the lowest priority number is given the highest priority. The priority scheduling algorithm can be preemptive or non-preemptive. In the non-preemptive priority scheduling algorithm, once a process is allocated the CPU, it will continue to


class PriorityQueue {
    constructor() {
        this.items = [];
    }

    enqueue(element, priority) {
        let qElement = { element, priority };
        let contain = false;

        for (let i = 0; i < this.items.length; i++) {
            if (this.items[i].priority > qElement.priority) {
                this.items.splice(i, 0, qElement);
                contain = true;
                break;
            }
        }

        if (!contain) this.items.push(qElement);
    }

    dequeue() {
        if (this.isEmpty()) return "Underflow";
        return this.items.shift();
    }

    front() {
        if (this.isEmpty()) return "No elements in Queue";
        return this.items[0];
    }

    rear() {
        if (this.isEmpty()) return "No elements in Queue";
        return this.items[this.items.length - 1];
    }

    isEmpty() {
        return this.items.length === 0;
    }

    printPQueue() {
        let str = "";
        for (let i = 0; i < this.items.length; i++) str += this.items[i].element + " ";
        return str;
    }
}

class PriorityScheduling {
    constructor() {
        this.queue = new PriorityQueue();
    }

    findWaitingTime(processes, n, bt, wt, priority) {
        let rt = [];
        for (let i = 0; i < n; i++) rt[i] = bt[i];
        let complete = 0;
        let t = 0;
        let minm = Number.MAX_VALUE;
        let shortest = 0;
        let finish_time;
        let found = false;
        let check = false;

        while (complete !== n) {
            for (let j = 0; j < n; j++) {
                if (processes[j] <= t && rt[j] < minm && rt[j] > 0) {
                    minm = rt[j];
                    shortest = j;
                    found = true;
                }
            }

            if (!found) {
                t++;
                continue;
            }

            rt[shortest]--;
            minm = rt[shortest];
            if (minm === 0) minm = Number.MAX_VALUE;

            if (rt[shortest] === 0) {
                complete++;
                found = false;
                finish_time = t + 1;
                wt[shortest] = finish_time - bt[shortest] - processes[shortest];
                if (wt[shortest] < 0) wt[shortest] = 0;
            }

            t++;
        }
    }

    findTurnAroundTime(processes, n, bt, wt, tat) {
        for (let i = 0; i < n; i++) tat[i] = bt[i] + wt[i];
    }
}

// Implementation
const processes = [1, 2, 3];
const n = processes.length;
const burst_time = [6, 8, 7];
const priority = [2, 1, 3];
const ps = new PriorityScheduling();
const wt = [];
const tat = [];
ps.findWaitingTime(processes, n, burst_time, wt, priority);
ps.findTurnAroundTime(processes, n, burst_time, wt, tat);
let total_wt = 0,
    total_tat = 0;
for (let i = 0; i < n; i++) {
    total_wt += wt[i];
    total_tat += tat[i];
}
const avg_wt = total_wt / n;
const avg_tat = total_tat / n;
console.log(`Average waiting time: ${avg_wt}`);
console.log(`Average turnaround time: ${avg_tat}`);

File: Operating_Systems/CPU-Scheduling/PriorityScheduling.py
# implement non-preemptive priority scheduling
# non-preemptive priority scheduling is a scheduling algorithm where the process with the highest priority is selected for execution. If two processes have the same priority, the process that arrived first is selected. The algorithm is non-preemptive, meaning that once a process is selected, it runs to completion without interruption.


def non_preemptive_priority_scheduling(processes):
    # Sort by Arrival Time, then Priority
    processes.sort(key=lambda x: (x['arrival'], x['priority']))
    
    # Initialize variables
    time = 0
    completed = 0
    n = len(processes)
    ready_queue = []
    results = []

    # Add a "remaining_time" field to track remaining burst time
    for p in processes:
        p['remaining_time'] = p['burst']
    
    while completed < n:
        # Add processes that have arrived to the ready queue
        for p in processes:
            if p['arrival'] <= time and p not in ready_queue and p['remaining_time'] > 0:
                ready_queue.append(p)
        
        # Sort ready queue by priority
        ready_queue.sort(key=lambda x: x['priority'], reverse=True)
        
        if ready_queue:
            # Pick the process with the highest priority
            current_process = ready_queue[0]
            time += current_process['remaining_time']
            current_process['completion'] = time
            current_process['remaining_time'] = 0
            completed += 1
            ready_queue.remove(current_process)
        else:
            # If no process is ready, move time forward
            time += 1

    # Calculate Turnaround Time (TAT) and Waiting Time (WT)
    for p in processes:
        p['tat'] = p['completion'] - p['arrival']
        p['wt'] = p['tat'] - p['burst']
        results.append(p)

    return results


# Example Processes
processes = [
    {'name': 'P1', 'arrival': 0, 'burst': 7, 'priority': 3},
    {'name': 'P2', 'arrival': 2, 'burst': 4, 'priority': 1},
    {'name': 'P3', 'arrival': 4, 'burst': 1, 'priority': 4},
    {'name': 'P4', 'arrival': 5, 'burst': 4, 'priority': 2}
]

# Run Non-Preemptive Priority Scheduling
results = non_preemptive_priority_scheduling(processes)

# Print Results
print("Process | Arrival | Burst | Priority | Completion | Turnaround | Waiting")
for p in results:
    print(f"{p['name']:>7} | {p['arrival']:>7} | {p['burst']:>5} | {p['priority']:>8} | {p['completion']:>10} | {p['tat']:>10} | {p['wt']:>7}")


# pre-emptive priority scheduling
# pre-emptive priority scheduling is a scheduling algorithm where the process with the highest priority is selected for execution. If two processes have the same priority, the process that arrived first is selected. The algorithm is pre-emptive, meaning that a process can be interrupted and replaced by a higher priority process.

def preemptive_priority_scheduling(processes):
    # Sort by Arrival Time, then Priority
    processes.sort(key=lambda x: (x['arrival'], x['priority']))
    
    # Initialize variables
    time = 0
    completed = 0
    n = len(processes)
    ready_queue = []
    results = []

    # Add a "remaining_time" field to track remaining burst time
    for p in processes:
        p['remaining_time'] = p['burst']
    
    while completed < n:
        # Add processes that have arrived to the ready queue
        for p in processes:
            if p['arrival'] <= time and p not in ready_queue and p['remaining_time'] > 0:
                ready_queue.append(p)
        
        # Sort ready queue by priority
        ready_queue.sort(key=lambda x: x['priority'], reverse=True)
        
        if ready_queue:
            # Pick the process with the highest priority
            current_process = ready_queue[0]
            current_process['remaining_time'] -= 1
            time += 1
            
            # If the process is complete
            if current_process['remaining_time'] == 0:
                current_process['completion'] = time
                completed += 1
                ready_queue.remove(current_process)
        else:
            # If no process is ready, move time forward
            time += 1

    # Calculate Turnaround Time (TAT) and Waiting Time (WT)
    for p in processes:
        p['tat'] = p['completion'] - p['arrival']
        p['wt'] = p['tat'] - p['burst']
        results.append(p)

    return results


# Example Processes
processes = [
    {'name': 'P1', 'arrival': 0, 'burst': 7, 'priority': 3},
    {'name': 'P2', 'arrival': 2, 'burst': 4, 'priority': 1},
    {'name': 'P3', 'arrival': 4, 'burst': 1, 'priority': 4},
    {'name': 'P4', 'arrival': 5, 'burst': 4, 'priority': 2}
]

# Run Preemptive Priority Scheduling
results = preemptive_priority_scheduling(processes)

# Print Results
print("Process | Arrival | Burst | Priority | Completion | Turnaround | Waiting")
for p in results:
    print(f"{p['name']:>7} | {p['arrival']:>7} | {p['burst']:>5} | {p['priority']:>8} | {p['completion']:>10} | {p['tat']:>10} | {p['wt']:>7}")

# Output: